### PR TITLE
perf: cache `Available` results to reduce network io

### PR DESCRIPTION
**Pull request by aooohan**
- Wrap the Available hook in the plugin with a layer of cache so that the cache can be used regardless of `vfox` calls or calls within the plugin.
- Allows configurable cache duration, defaults to 12 hours.

see https://github.com/version-fox/vfox/issues/227#issuecomment-2074205042

### CODE REVIEW

**Comment by Chance-fyi** [comment]
- We should be able to refresh the cache manually.
- Should the cache be invalidated for the following operation?

  ```bash
  vfox config cache.availableHookDuration 12h
  # Cache and sleep
  vfox config cache.availableHookDuration 1s
  ```

**Comment by aooohan** [comment]
> * We should be able to refresh the cache manually.

Agreed, but this can be enhanced later, for now it can be done with the `config` command.

> * Should the cache be invalidated for the following operation?


The cache is invalidated only in the following cases:
```shell 
vfox config cache.availableHookDuration 0

vfox config cache.availableHookDuration -1
```

In addition, there seems to be an issue with `config` command. I cannot directly assign `1s` to `availableHookDuration` that is a field of type `time.Duration` by command.

**Comment by bytemain** [comment]
and we also can provide a cache library for the plugins, let them can store some data.


```lua main.lua
let cache = require("vfox/cache")

local key = "nodejs-index-json"
local hit = cache.has(key)
if (!hit) then
  local data = -- fetch --
  cache.set(key, data, 12 * 60 * 60 * 1000)
end

```

```

**Comment by aooohan** [comment]
> and we also can provide a cache library for the plugins, let them can store some data.

Sure.  I have this plan.

**Comment by Chance-fyi** [comment]
> In addition, there seems to be an issue with config command. I cannot directly assign 1s to availableHookDuration that is a field of type time.Duration by command.

I've added support for `time.Duration` type, but haven't pushed yet.

image

I noticed that YAML parsing fails if there's no type.

image

**Comment by aooohan** [comment]
> I noticed that YAML parsing fails if there's no type.

Oops, thanks for point out. Let me think about how to deal with it.

**Comment by jan-bar** [comment]
> I noticed that YAML parsing fails if there's no type.

You can try adding the following code

```go
func (c *Cache) UnmarshalYAML(node *yaml.Node) error {
	var data map[string]any
	err := node.Decode(&data)
	if err != nil {
		return err
	}

	if v, ok := data["availableHookDuration"]; ok {
		switch va := v.(type) {
		case int:
			c.AvailableHookDuration = time.Duration(va)
		case string:
			c.AvailableHookDuration, err = time.ParseDuration(va)
			if err != nil {
				return err
			}
		}
	}
	return nil
}
```

Or

```go
type Cache struct {
	AvailableHookDuration VfoxDuration `yaml:"availableHookDuration"` // Available hook result cache time
}

type VfoxDuration time.Duration

func (c *VfoxDuration) UnmarshalYAML(node *yaml.Node) error {
	var data any
	err := node.Decode(&data)
	if err != nil {
		return err
	}

	switch va := data.(type) {
	case int:
		*c = VfoxDuration(va)
	case string:
		pd, err := time.ParseDuration(va)
		if err != nil {
			return err
		}
		*c = VfoxDuration(pd)
	}
	return nil
}
```

The second option adds a new type
The first option also needs to modify this part of the code when adding fields to the Cache later

**Comment by aooohan** [comment]
> The second option adds a new type

I took your second option. ; )

**Review by bytemain — APPROVED** [review]

File: internal/plugin.go [hunk 1]
```diff
@@ -526,6 +528,83 @@ func NewLuaPlugin(pluginDirPath string, manager *Manager) (*LuaPlugin, error) {
 	if source.Name == "" {
 		return nil, fmt.Errorf("no plugin name provided")
 	}
+
+	// wrap Available hook with Cache.
+	if source.HasFunction("Available") {
+		targetHook := PLUGIN.RawGetString("Available")
+		source.pluginObj.RawSetString("Available", vm.Instance.NewFunction(func(L *lua.LState) int {
```
Comment (bytemain) [inline internal/plugin.go:535]: 看起来可以把这个 hook 的能力做到 LuaVM 里，

一个初始的想法是这样的：

vm.DecorateFunction(func, func () int {})

Comment (aooohan) [reply internal/plugin.go:535]: 好主意! 可以放到下一版本去做. 目前先保证功能优先.

**Comment by aooohan** [comment]
> I've added support for `time.Duration` type, but haven't pushed yet.

Hi, Can you push your changes?

**Pull request merged**